File: MATTHEW/leetcode/AIcoding/4.GPT-4o/python3_solutions/872-leaf-similar-trees.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def leafSimilar(root1: TreeNode, root2: TreeNode) -> bool:
    def getLeaves(node):
        if not node:
            return []
        if not node.left and not node.right:
            return [node.val]
        return getLeaves(node.left) + getLeaves(node.right)

    return getLeaves(root1) == getLeaves(root2)

File: MATTHEW/leetcode/AIcoding/4.GPT-4o/rust_solutions/872-leaf-similar-trees.rs
use std::rc::Rc;
use std::cell::RefCell;

#[derive(Debug)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

pub fn leaf_similar(root1: Option<Rc<RefCell<TreeNode>>>, root2: Option<Rc<RefCell<TreeNode>>>) -> bool {
    fn get_leaves(root: Option<Rc<RefCell<TreeNode>>>, leaves: &mut Vec<i32>) {
        if let Some(node) = root {
            let node = node.borrow();
            if node.left.is_none() && node.right.is_none() {
                leaves.push(node.val);
            } else {
                get_leaves(node.left.clone(), leaves);
                get_leaves(node.right.clone(), leaves);
            }
        }
    }

    let mut leaves1 = Vec::new();
    let mut leaves2 = Vec::new();

    get_leaves(root1, &mut leaves1);
    get_leaves(root2, &mut leaves2);

    leaves1 == leaves2
}